sycl/source/detail/global_handler.hpp

```diff
@@ -0,0 +1,77 @@
+//==--------- global_handler.hpp --- Global objects handler ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/spinlock.hpp>
+#include <CL/sycl/detail/util.hpp>
+
+#include <memory>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+class platform_impl;
+class Scheduler;
+class ProgramManager;
+class Sync;
+class plugin;
+
+using PlatformImplPtr = std::shared_ptr<platform_impl>;
+
+/// Wrapper class for global data structures with non-trivial destructors.
+///
+/// As user code can call SYCL Runtime functions from destructor of global
+/// objects, it is not safe for the runtime library to have global objects with
+/// non-trivial destructors. Such destructors can be called any time after
+/// exiting main, which may result in user application crashes. Instead,
+/// complex global objects must be wrapped into GlobalHandler. Its instance
+/// is stored on heap, and deallocated when the runtime library is being
+/// unloaded.
+///
+/// There's no need to store trivial globals here, as no code for their
+/// construction or destruction is generated anyway.
+class GlobalHandler {
+public:
+  /// \return a reference to a GlobalHandler singleton instance. Memory for
+  /// storing objects is allocated on first call. The reference is valid as long
+  /// as runtime library is loaded (i.e. untill `DllMain` or
+  /// `__attribute__((destructor))` is called).
+  static GlobalHandler &instance();
+
+  GlobalHandler(const GlobalHandler &) = delete;
+  GlobalHandler(GlobalHandler &&) = delete;
+
+  Scheduler &getScheduler();
+  ProgramManager &getProgramManager();
+  Sync &getSync();
+  std::vector<PlatformImplPtr> &getPlatformCache();
+  std::mutex &getPlatformMapMutex();
+  std::mutex &getFilterMutex();
+  std::vector<plugin> &getPlugins();
+
+private:
+  friend void shutdown();
+  // Constructor and destructor are declared out-of-line to allow incomplete
+  // types as template arguments to unique_ptr.
+  GlobalHandler();
+  ~GlobalHandler();
+
+  SpinLock &getFieldsLock() { return MFieldsLock; }
+
+  SpinLock MFieldsLock;
+
+  std::unique_ptr<Scheduler> MScheduler;
```

Comment:
Can we avoid truly knowing/enumerating all the objects handled in this wrapper, and just make it be a custom heap where arbitrary objects can be dealt with overloaded new/delete?

Reply:
@smaslov-intel I was thinking about it. If we go with some custom memory space, we'd need some mechanism to call destructors on that memory. We'd either need to somehow store custom deleters and call them upon shutdown, or implement some other logic, where object is responsible for its destruction. Both look quite complicated solutions. Also, it's not clear how to access these objects. Use names and store pointers in a map? So, I decided to keep global wrapper with unique pointers for now. Adding a new global object shouldn't be that hard, and I don't expect us to add a new global object every other Tuesday as they're code smell anyway.

Reply:
> Adding a new global object shouldn't be that hard

But it is badly violating OOP encapsulation principle, I feel.

Also it makes it nearly impossible to use this global handler in the plugins (standalone parts of SYCL RT).
And then anything else we use for plugins would compete with 65535 destruction priority used for this data.

Having said that I admit this is definitely an improvement and we should go with it until something better is developed.